### PR TITLE
Use AudioStream and VideoStream (which inherits MediaStream)

### PR DESCRIPTION
**Pull request by ChristopheI**
Creating this PR to track progress on this implementation and discuss.

The goals if to centralize objects, properties, method and logic in **AudioStream** and **VideoStream** objects instead to have all this dispatch in the **RTPSession** class.

**AudioStream** and **VideoStream** inherits **MediaStream** to share common stuff.

It will permit to reduce size of  **RTPSession** class and a cleaner code. It will also permits later to implement support of multiple **AudioStream** and/or **VideoStream** 

The signature of  **RTPSession**  must be changed a little but in a very straightforward way. For example:

- RTPSession.AudioLocalTrack => RTPSession.AudioStream.LocalTrack
- RTPSession.VideoRemoteTrack => RTPSession.VideoStream.RemoteTrack
- RTPSession.AddBuffer(SDPMediaTypesEnum.audio, timeSpan) => RTPSession.AudtioStream.AddBuffer(timeSpan)
- etc ...

### Content of new classes

**MediaStream:**

- [x] event OnTimeout
- [x] event OnSendReport
- [x] event OnRtpPacketReceived
- [x] event OnRtpEvent
- [x] event OnReceiveReport
- [x] properties MediaType
- [x] properties LocalTrack
- [x] properties RemoteTrack
- [x] properties RtcpSession
- [x] properties DestinationEndPoint
- [x] properties ControlDestinationEndPoint
- [x] properties IsClosed
- [x] properties RemoteRtpEventPayloadID
- [x] SecureContext stuff
- [x] OrderBuffer stuff
- [x] method SendRtpPacket
- [x] method SendRtpRaw
- [x] method SendRtcpReport
- [x] method SendRtcpReport
- [x] method SendRtcpFeedback
- [x] method GetSendingFormat
- [x] method ProcessHeaderExtensions
- [x] method SetDestination
- [x] method CreateRtcpSession
- [x] method OnReceiveRTPPacket

**AudioStream:**

- [x] event OnAudioFormatsNegotiated
- [x] method CheckAudioFormatsNegotiation
- [x] method HasAudio
- [x] method SendAudio
- [x] method SendAudioFrame
- [x] method SendDtmfEvent
- [x] method SendDtmf
- [x] method OnReceiveRTPPacket

**VideoStream:**

- [x] event OnVideoFormatsNegotiated
- [x] event OnVideoFrameReceived
- [x] method CheckVideoFormatsNegotiation
- [x] method HasVideo
- [x] method ProcessVideoRtpFrame
- [x] method SendJpegFrame
- [x] method SendH264Frame
- [x] method SendH264Nal
- [x] method SendVp8Frame
- [x] method SendVideo
- [x] method OnReceiveRTPPacket

### Current class

**RTPSession:** (To check / clean / update)

- [x] method OnReceive
- [x] RTP Channel stuff

### Update samples 

- [ ] AzureExamples
- [ ] SctpClientTestConsole
- [ ] sipcmdline
- [ ] SipExamples/AndroidSIPGetStarted
- [ ] SipExamples/AsteriskIce
- [ ] SipExamples/AttendedTransfer
- [ ] SipExamples/CallHoldAndTransfer
- [ ] SipExamples/CustomAudioCodec
- [ ] SipExamples/GetStarted
- [ ] SipExamples/GetStartedNetFramework
- [ ] SipExamples/GetStartedPortAudio
- [ ] SipExamples/GetStartedVideo
- [ ] SipExamples/GetStartedWebSocket
- [ ] SipExamples/iOSSIPGetStarted
- [ ] SipExamples/NotifierClient
- [ ] SipExamples/PlaySounds
- [ ] SipExamples/RecordCall
- [ ] SipExamples/RecordIncomingCall
- [ ] SipExamples/SendDtmf
- [ ] SipExamples/SendMusic
- [ ] SipExamples/SIPCallResampleAudio
- [ ] SipExamples/SIPCallServer
- [ ] SipExamples/SIPProxy
- [ ] SipExamples/UserAgentClient
- [ ] SipExamples/UserAgentRegister
- [ ] SipExamples/UserAgentServer
- [ ] SipExamples/VideoPhoneCmdLine
- [ ] SIPScenarios/AttendedTransferScenario
- [ ] SIPScenarios/BlindTransferScenario
- [ ] SIPScenarios/LoadTestScenario
- [ ] SIPScenarios/OnHoldScenario
- [ ] Softphone
- [ ] StunServer
- [x] webrtccmdline
- [x] WebRTCExamples/FfmpegToWebRTC
- [ ] WebRTCExamples/JanusWebRTCStream
- [ ] WebRTCExamples/sipjs
- [ ] WebRTCExamples/SIPToWebRtcBridge
- [ ] WebRTCExamples/SIPToWebRtcBridgeVideo
- [x] WebRTCExamples/WebRTCAspNetMvc
- [x] WebRTCExamples/WebRTCClient
- [ ] WebRTCExamples/WebRTCDaemon
- [ ] WebRTCExamples/WebRTCEchoClient
- [x] WebRTCExamples/WebRTCEchoServer
- [ ] WebRTCExamples/WebRTCGetStarted
- [x] WebRTCExamples/WebRTCGetStartedDataChannel
- [x] WebRTCExamples/WebRTCMp4Source
- [ ] WebRTCExamples/WebRTCNoSignalling
- [x] WebRTCExamples/WebRTCReceiveAudio
- [x] WebRTCExamples/WebRTCReceiver
- [x] WebRTCExamples/WebRTCSendAudio
- [x] WebRTCExamples/WebRTCTestPatternRest
- [ ] WebRTCExamples/WebRTCTestPatternServer
- [x] WebRTCExamples/WebRTCtoFfplay
- [ ] WebRTCExamples/WebRTCWebcamServer

### CODE REVIEW

**Comment by Cheroukee** [comment]
Since you asked me too about the security context stuff, I will add my 2¢ : 

Moving the video and audio specific stuff away from the RTP Session to me sounds like a great idea since the RTP Session class is pretty big (in my opinion).

The PR is pretty huge to review since the RTP Session has been modified so much. If possible, could you split the `MediaStream` file into one file per class

For your goal which seems to be to enable the use of multiple video streams/multiple audio streams inside a single WebRTC session, I feel like this could be achieved by having two different connections instead without the requiring these modifications. Is there something holding you back from doing this ?

Otherwise seems like doing multiple video/audio `m=<...>` lines is possible. I'll review it more in depth tomorrow

**Comment by ChristopheI** [comment]
> The PR is pretty huge to review since the RTP Session has been modified so much. If possible, could you split the `MediaStream` file into one file per class

Sure.
 
> For your goal which seems to be to enable the use of multiple video streams/multiple audio streams inside a single WebRTC session, I feel like this could be achieved by having two different connections instead without the requiring these modifications. Is there something holding you back from doing this ?

Sure it could be possible to create several PeerConnection but I have already some use case (web cam + desktop sharing for example) which needs to have at least 2 video streams in one connection
 
> Otherwise seems like doing multiple video/audio `m=<...>` lines is possible. I'll review it more in depth tomorrow

Thx

File: src/net/RTP/RTPSession.cs [hunk 1]
```diff
@@ -1401,63 +1365,30 @@ protected virtual RTPChannel CreateRtpChannel(SDPMediaTypesEnum mediaType)
...
-            ProtectRtpPacket unprotectRtp,
-            ProtectRtpPacket protectRtcp,
-            ProtectRtpPacket unprotectRtcp)
+        protected void AddRtpChannel(SDPMediaTypesEnum mediaType, RTPChannel rtpChannel)
         {
-            foreach (var mediaType in mediaTypes)
+            if (rtpSessionConfig.IsMediaMultiplexed)
             {
-                SetSecurityContext(mediaType, protectRtp, unprotectRtp, protectRtcp, unprotectRtcp);
+                AudioStream.AddRtpChannel(rtpChannel);
```
[the remark this answers is not in the document]
Comment (ChristopheI) [reply src/net/RTP/RTPSession.cs]: I change the way to manage RTPChannel. 
Could you check if it's ok for you ?

**Comment by sipsorcery** [comment]
Thanks a lot for putting the work into this PR. It's a big job.

I agree that it makes a lot of sense to refactor RTPSession. And separating the base media stream, audio and video logc makes perfect sense.

One suggestion would be, instead of:

 - RTPSession.AudioLocalTrack => RTPSession.AudioStream.LocalTrack
 - RTPSession.VideoRemoteTrack => RTPSession.VideoStream.RemoteTrack
 - etc.

Use both mechanisms so as not to break all existing applications. For example leave the `RTPSession.AudioLocalTrack` property on `RTPSession` but wire it up to return `RTPSession.AudioStream.LocalTrack`. Nothing annoys devs more than chaning the public API surface of a library. If it can be avoided, even if it results in slightly dirty code, we should aim to do so.

Overall though big concept ACK from me. I'll give the PR more of a spin once you think it's ready for more in depth testing. Or is that already the case?

**Comment by ChristopheI** [comment]
@sipsorcery ,

AudioLocalTrack, AudioRemoteTrack, AudioRtcpSession, VideoLocalTrack, VideoRemoteTrack, VideoRtcpSession are now again available directly

I have merge also last commit from the master branch.

I think I made all necessaries updates so if you could take a deeper look to this PR it will be great.

Thanks.

**Comment by sipsorcery** [comment]
@ChristopheI the new AudioStream and VideoStream classes look good.

What's happening with the DTLS configuration? Does this PR revert @Cheroukee's PR that allowed different DTLS contexts for audio and video? It seems that with this PR only a single DTLS context can be set up and assigned to the AudioStream?

**Comment by ChristopheI** [comment]
> What's happening with the DTLS configuration? Does this PR revert @Cheroukee's PR that allowed different DTLS contexts for audio and video? It seems that with this PR only a single DTLS context can be set up and assigned to the AudioStream?

I badly managed this part in RTCPeerConnection. I just fix this part.

I have added also public properties / methods in RTPSession

**Comment by sipsorcery** [comment]
@ChristopheI I tested out this PR with 3 WebRTC examples and one SIP one. All seemed to be in good order (FFmpeg dll's are missing but that's not related to this PR) except for [WebRTCMp4Source](https://github.com/sipsorcery-org/sipsorcery/tree/master/examples/WebRTCExamples/WebRTCMp4Source) which crashed. Would you be able to give that example a go and see what you think? I suspect it's related to other changes and not this PR but you're probably in a better position to say.

**Comment by sipsorcery** [comment]
@ChristopheI thanks for the latest changes and apologies for the weekly delays in reviews.

Looks good now, I'll go ahad and merge.

**Pull request merged**